core/storage/blog/gae_models.py

```diff
@@ -0,0 +1,424 @@
+    thumbnail_filename = datastore_services.StringProperty(indexed=True)
+    # Time when the blog post model was last published. Value will be None
+    # if the blog has never been published.
+    published_on = (
+        datastore_services.DateTimeProperty(indexed=True))
+
+    @staticmethod
+    def get_deletion_policy():
+        """Model contains data to pseudonymize corresponding to a user:
+        author_id field.
+        """
+        return base_models.DELETION_POLICY.LOCALLY_PSEUDONYMIZE
+
+    @classmethod
+    def has_reference_to_user_id(cls, user_id):
+        """Check whether BlogPostRightsModel references user.
+
+        Args:
+            user_id: str. The ID of the user whose data should be checked.
+
+        Returns:
+            bool. Whether any models refer to the given user ID.
+        """
+        return cls.query(
+            cls.author_id == user_id
+        ).get(keys_only=True) is not None
+
+    @staticmethod
+    def get_model_association_to_user():
+        """Model is exported as multiple instances per user since there can
+        be multiple blog post models relevant to a user.
+        """
+        return base_models.MODEL_ASSOCIATION_TO_USER.MULTIPLE_INSTANCES_PER_USER
+
+    @classmethod
+    def get_export_policy(cls):
+        """Model contains data corresponding to a user to export."""
+        return dict(super(BlogPostModel, cls).get_export_policy(), **{
+            # We do not export the author_id because we should not
+            # export internal user ids.
+            'author_id': base_models.EXPORT_POLICY.NOT_APPLICABLE,
+            'title': base_models.EXPORT_POLICY.EXPORTED,
+            'content': base_models.EXPORT_POLICY.EXPORTED,
+            'url_fragment': base_models.EXPORT_POLICY.EXPORTED,
+            'tags': base_models.EXPORT_POLICY.EXPORTED,
+            'thumbnail_filename': base_models.EXPORT_POLICY.EXPORTED,
+            'published_on': base_models.EXPORT_POLICY.EXPORTED,
+        })
+
+    @classmethod
+    def generate_new_blog_post_id(cls):
+        """Generates a new blog post ID which is unique and is in the form of
+        random hash of 12 chars.
+
+        Returns:
+            str. A blog post ID that is different from the IDs of all
+            the existing blog posts.
+
+        Raises:
+            Exception. There were too many collisions with existing blog post
+                IDs when attempting to generate a new blog post ID.
+        """
+
+        for _ in python_utils.RANGE(base_models.MAX_RETRIES):
+            blog_post_id = utils.convert_to_hash(
+                python_utils.UNICODE(
+                    utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+            if not cls.get_by_id(blog_post_id):
+                return blog_post_id
+        raise Exception(
+            'New blog post id generator is producing too many collisions.')
+
+    @classmethod
+    def create(cls, blog_post_id, author_id):
```

Comment:
The base models  that are not versioned in our code base use create method in storage layer rather than in domain. eg: feedback, AppFeedbackReportModel, TaskEntryModel ( in improvements). 
We will call this method in the service layer to generate the first instance of the blog post. This is so that a blog post is created even if the user does not save it as draft or publishes it.

Reply:
But we should create the domain object first and can only save it to model/datastore once it's valid (that is why we have the domain layer)

by "valid" i mean the schema is valid and passes the BlogPost.validate method.

Reply:
The created object is the valid one. We will have 2 types of validations, this model will fail validation in strict mode.